### DISCUSSION TITLE
apm-toolkit-meter doesnt work

### DISCUSSION THREAD

**Discussion by 9029HIME** [apache/skywalking#11319] [Q&A]
Version Info:
```
skywalking-jdk:jdk17
skywalking-apm:9.4.0
java-agent:9.0.0
application jdk:jdk8
apm-toolkit-meter dependency:9.0.0
```

I have put a custom meter file in **apache-skywalking-apm-bin/config/meter-analyzer-config/business_success_count.yaml** like this:
```
expSuffix: instance(['service'], ['instance'], Layer.GENERAL)
metricPrefix: meter
metricsRules:
  - name: business_success_count
    exp: business_success_count.increase("PT1M")
```

and I have modifed **application.yaml** like this:
```
meterAnalyzerActiveFiles: ${SW_METER_ANALYZER_ACTIVE_FILES:datasource,threadpool,satellite,go-runtime,python-runtime,business_success_count}
````

after starting skywalking, I found that there was a table named **meter_business_success_count** in the databases, as I inserted one record into this table manually and that showed on UI:

![image](https://github.com/apache/skywalking/assets/48977889/ab30e84a-5387-4c8b-b6b5-51e834f93841)

![image](https://github.com/apache/skywalking/assets/48977889/f30c7e3d-9791-4d24-a57d-88024f6a9fd1)

But After executing the following code, there was no any record was inserted into table:
```
@GetMapping("/test")
    public RestResult test(){
        String name = "meter_business_success_count";
        Counter counter = MeterFactory.counter(new MeterId (name,MeterId.MeterType.COUNTER))
                .mode(Counter.Mode.INCREMENT).build();
        counter.increment(10D);
        logger.info("after：{}:{}", counter.getName(),counter.get());
        return RestResult.success();
    }
```

I am sure that **apm-toolkit-meter-activation-9.0.0** was loaded into JVM according to following logs:
```
main AgentClassLoader : D:\apache-skywalking-java-agent-9.0.0\skywalking-agent\activations\apm-toolkit-meter-activation-9.0.0.jar loaded.
```
```
find skywalking plugin define in jar:file:D:\apache-skywalking-java-agent-9.0.0\skywalking-agent\activations\apm-toolkit-meter-activation-9.0.0.jar!/skywalking-plugin.def
```

I have tried to find some relevant documents and issues, but I have no idea about clearing my question. Is there anything I have ignored?
Thank you for your analysis and help.

**Comment by wu-sheng** [comment]
You are creating a counter repeatedly? You should keep created counter once and use it later.

**Reply by 9029HIME** [reply]
Thank you for your analysis.

I could be wrong, but I just changed my code structure to this:
```
private Counter counter = MeterFactory.counter(new MeterId("meter_business_success_count",MeterId.MeterType.COUNTER))
            .mode(Counter.Mode.INCREMENT).build();

@GetMapping("/test")
public RestResult test(){
    counter.increment(10D);
    logger.info("after：{}:{}", counter.getName(),counter.get());
    return RestResult.success();
}
```
But after executing for 15 times, it still didnt work while other metric like **service_instance_sla** was uploaded to sw server and recalculated:

![image](https://github.com/apache/skywalking/assets/48977889/59b8e581-cea3-4c3f-ac75-d3a62d960a61)

![image](https://github.com/apache/skywalking/assets/48977889/726a913a-9088-493f-a73b-f24fa28e79cc)

![image](https://github.com/apache/skywalking/assets/48977889/dda486ac-0cd5-4d9f-9f5b-4b506633fd75)

Maybe it is not due to repetitive creation?

**Reply by wu-sheng** [reply]
Read this, https://github.com/apache/skywalking-java/blob/bc7447a2c2de1c6f11ba8bd6f35dd64a2c474746/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/meter/MeterService.java#L64

If you created repeated, the meaningful ref would be lost inevitably.

**Reply by 9029HIME** [reply]
Thank you so much for the link you found for me, I just changed my code to that create counter only once, as I mentioned in my previous reply.

But it still doesnt take effect after requesting **/test** for many times.

**Reply by wu-sheng** [reply]
Then you have to debug on your own. I don't have more to share.

**Reply by 9029HIME** [reply]
Many thanks
